pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -52,138 +69,430 @@ some differences:
-		# Meanwhile, as a workaround, delete probes manually.
-		# See: https://github.com/iovisor/gobpf/issues/223
-		echo "-:pfree_uts_ns" >> /sys/kernel/debug/tracing/kprobe_events 2>/dev/null || true
-		echo "-:pcap_capable" >> /sys/kernel/debug/tracing/kprobe_events 2>/dev/null || true
-
-		# Remove the opened files limit to avoid getting this error:
-		# error while loading "tracepoint/raw_syscalls/sys_enter" (too many open files):
-		# Indeed, traceloop creates a lof of tracers which each has maps and events:
-		# https://github.com/kinvolk/traceloop/blob/6f4efc6fca46d92c75f4ec4e6c6e1d829bdeaddf/bpf/straceback-guess-bpf.h#L27-L28
-		# So, this can generate a lof ot opened files.
-		ulimit -n hard
-
-		rm -f /run/traceloop.socket
-		exec /bin/traceloop k8s
-	`)
-	t.cmd.Stdout = os.Stdout
-	t.cmd.Stderr = os.Stderr
-	err := t.cmd.Start()
-	if err != nil {
-		trace.Status.OperationError = fmt.Sprintf("Failed to start: %s", err)
+	if trace.Spec.OutputMode != gadgetv1alpha1.TraceOutputModeStatus {
+		trace.Status.OperationError = fmt.Sprintf("\"start\" operation can only be used with %q trace while %q was given", gadgetv1alpha1.TraceOutputModeStatus, trace.Spec.OutputMode)
+
 		return
 	}
+
+	if t.started {
+		trace.Status.State = gadgetv1alpha1.TraceStateStarted
+		return
+	}
+
+	// Having this backlink is mandatory for delete operation.
+	t.trace = trace
+
+	// The output will contain an array of types.TraceloopInfo.
+	// So, to avoid problem, we initialize it to be a JSON array.
+	trace.Status.Output = "[]"
+	t.containerIDs = make(map[string]uint64, 0)
+
+	genKey := func(container *containercollection.Container) string {
+		return container.Namespace + "/" + container.Podname
+	}
+
+	attachContainerFunc := func(container *containercollection.Container) error {
+		containerID := container.ID
+		mntNsID := container.Mntns
+		key := genKey(container)
+
+		traceUnique.Lock()
+		err := traceUnique.tracer.Attach(containerID, mntNsID)
+		traceUnique.Unlock()
+		if err != nil {
+			log.Errorf("failed to attach tracer: %s", err)
+
+			return err
+		}
+
+		t.containerIDs[containerID] = mntNsID
+		log.Debugf("tracer attached for %q (%d)", key, mntNsID)
+
+		// There is no client with local-gadget, so we can quit right now.
+		if t.client == nil {
+			return nil
+		}
+
+		var infos []types.TraceloopInfo
+		err = json.Unmarshal([]byte(trace.Status.Output), &infos)
+		if err != nil {
+			log.Errorf("failed to unmarshal output: %s", err)
+
+			return err
+		}
+
+		infos = append(infos, types.TraceloopInfo{
+			Namespace:     container.Namespace,
+			Podname:       container.Podname,
+			Containername: container.Name,
+			ContainerID:   containerID,
+		})
+
+		output, err := json.Marshal(infos)
+		if err != nil {
+			log.Errorf("failed to marshal infos: %s", err)
+
+			return err
+		}
+
+		traceBeforePatch := trace.DeepCopy()
+		trace.Status.Output = string(output)
```

Comment:
What's the character limit on `trace.Status.Output` and how many containers can we store there?

Reply:
Each Kubernetes resource cannot go over 1MB: https://github.com/etcd-io/etcd/issues/9925

Each trace only handle containers on a local node.

Kubernetes has a default maximum of [110 Pods per node](https://kubernetes.io/docs/setup/best-practices/cluster-large/). If all pods have a `pause` container and one sidecar, that's 330 containers per node.

330 containers * 64 characters for an id = 21KB.

So I think this is fine.